mgmt/src/models/internal/nat/table_extend.rs

```diff
@@ -0,0 +1,242 @@
+                // at the beginning of the function.
+                let new_length = prefix.length() + 1;
+                let mut new_address;
+                if prefix.as_address() == exclude.as_address() {
+                    // Exclusion prefix covers the first half of the prefix.
+                    // Here we need to update the address to keep the second half of the prefix.
+                    new_address = match prefix.as_address() {
+                        IpAddr::V4(addr) => {
+                            let Ok(exclude_size) = u32::try_from(exclude.size()) else {
+                                unreachable!(
+                                    "Exclude size too big ({}), bug in IpList",
+                                    exclude.size()
+                                )
+                            };
+                            IpAddr::V4(Ipv4Addr::from(addr.to_bits() + exclude_size))
+                        }
+                        IpAddr::V6(addr) => {
+                            IpAddr::V6(Ipv6Addr::from(addr.to_bits() + exclude.size()))
+                        }
+                        // Prefix cannot cover exclusion prefix of a different IP version
+                        _ => unreachable!(
+                            "Prefix and exclusion prefix are not of the same IP version"
+                        ),
+                    }
+                } else {
+                    // Exclusion prefix is the second half of the prefix; keep the first half.
+                    new_address = prefix.as_address();
+                }
+                let Ok(new_prefix) = Prefix::try_from((new_address, new_length)) else {
+                    unreachable!("Failed to create new prefix from ({new_address}, {new_length})");
+                };
+
+                clone.remove(prefix);
+                clone_not.remove(exclude);
+                clone.insert(new_prefix);
+            }
+        }
+    }
+    (clone, clone_not)
+}
+
+/// Optimize a [`Peering`] object:
+///
+/// - Optimize both [`VpcManifest`] objects
+#[must_use]
+pub fn optimize_peering(peering: &Peering) -> Peering {
+    // Collapse prefixes and exclusion prefixes
+    let mut clone = peering.clone();
+    for expose in &mut clone.local.exposes {
+        let (ips, nots) = optimize_expose(&expose.ips, &expose.nots);
+        expose.ips = ips;
+        expose.nots = nots;
+    }
+    for expose in &mut clone.remote.exposes {
+        let (as_range, not_as) = optimize_expose(&expose.as_range, &expose.not_as);
+        expose.as_range = as_range;
+        expose.not_as = not_as;
+    }
+    clone
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::models::internal::nat::tables::NatTables;
+    use iptrie::{Ipv4Prefix, Ipv6Prefix};
+    use net::vxlan::Vni;
+    use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+    use std::str::FromStr;
+
+    #[test]
+    fn test_fabric() {
```

Comment:
Fuzz testing would be really nice to validate the correctness of some of these optimizations.

Reply:
I suspect this optimisation will go away, or change, at some point. But other than that, I agree. I just haven't found the time to get into how fuzzing works, yet. I'll look into it at some point.